src/edu/psu/sweng500/team8/play/GameSession.java

```diff
@@ -56,39 +58,72 @@ public Board getGameBoard() {
 	 * @param cell
 	 * @param number
 	 */
-	public void enterNumber(Cell currentCell, int number) {
-		
-		/* keep track of the last action*/
-		SudokuAction sudokuAction = new SudokuAction(new CellGrid(board.getCellGrid()));
-		
-		boolean isDelete = false;
-		
-		if (number == 0) {
-			isDelete = true;			
-		}
-		
-		if (isDelete) {
-			currentCell.clearNumber();
-		} else {
-			currentCell.setNumber(number);	
-		}
-		
-		/* FIXME - PencilMarkManager updating/wiping out PencilMark number matching entered number */
-		if (!isDelete) {
-			for (int i = 0; i < 9; i++) {
-				for (int j = 0; j < 9; j++) {
-					Cell eachCell = this.getGameBoard().getCellGrid().getCell(i, j);
+	public void enterNumber(Cell currentCell, int number, boolean isHint) {
+
+		/* We don't have to clear empty cell */
+		if (0 != currentCell.getNumber() || 0 != number) {
+
+			/* TODO - this is a tricky part that may require refactoring later */
+			if (isHint) {
+				// Assuming hint will only populate empty cell 
+				currentCell.clearNumber();				
+			}
+			
+			/* keep track of the last action */
+			SudokuAction sudokuAction = new SudokuAction(new CellGrid(
+					board.getCellGrid()));
+
+			boolean isDelete = false;
+
+			if (number == 0) {
+				isDelete = true;
+			}
+
+			if (isDelete) {
+				currentCell.clearNumber();
+				currentCell.getPencilMarks().clear();
+			} else {
+				/*
+				 * TODO Verify : if a number is entered, the pencilMark numbers
+				 * are cleared?
+				 */
+				currentCell.getPencilMarks().clear();
+				currentCell.setNumber(number);
+			}
+
+			/* Delete PencilMark for the same row/column */

```

Comment:
What about blocks?

Just use getCellConstraints and iterate the cells in each one